### PR TITLE
feat(evaluate): serialize map and set

### LINKED ISSUE #24040: [Feature] Support passing Map/Set as props in component testing
If you currently pass a Map or a Set in the component testing props, you will get blank object. It would be quite helpful if this just worked. Restricting component props to just plain JSON is limiting and not always the case.

### PR DESCRIPTION
**Pull request by sand4rt**
closes: https://github.com/microsoft/playwright/issues/24040

### CODE REVIEW

**Comment by mxschmitt** [comment]
I'd do something like this instead which follows our existing serialisation infra, which would also work with circular objects, complex data types inside Map etc. This would be worth adding some tests.

```diff
diff --git a/packages/playwright-core/src/protocol/serializers.ts b/packages/playwright-core/src/protocol/serializers.ts
index 3e1f99a41..813ab22cb 100644
--- a/packages/playwright-core/src/protocol/serializers.ts
+++ b/packages/playwright-core/src/protocol/serializers.ts
@@ -76,9 +76,9 @@ function innerParseSerializedValue(value: SerializedValue, handles: any[] | unde
   if (value.r !== undefined)
     return new RegExp(value.r.p, value.r.f);
   if (value.m !== undefined)
-    return new Map(Object.entries(JSON.parse(value.m)));
+    return new Map(Object.entries(innerParseSerializedValue(value.m, handles, refs)));
   if (value.se !== undefined)
-    return new Set(JSON.parse(value.se));
+    return new Set(innerParseSerializedValue(value.se, handles, refs));
 
   if (value.a !== undefined) {
     const result: any[] = [];
@@ -150,9 +150,9 @@ function innerSerializeValue(value: any, handleSerializer: (value: any) => Handl
     return { s: `${error.name}: ${error.message}\n${error.stack}` };
   }
   if (isMap(value))
-    return { m: JSON.stringify(Object.fromEntries(value)) };
+    return { m: innerSerializeValue(Object.fromEntries(value), handleSerializer, visitorInfo) };
   if (isSet(value))
-    return { se: JSON.stringify(Array.from(value)) };
+    return { se: innerSerializeValue(Array.from(value), handleSerializer, visitorInfo) };
   if (isDate(value))
     return { d: value.toJSON() };
   if (isURL(value))
diff --git a/packages/playwright-core/src/protocol/validator.ts b/packages/playwright-core/src/protocol/validator.ts
index 71f1ffbd0..c05a349a5 100644
--- a/packages/playwright-core/src/protocol/validator.ts
+++ b/packages/playwright-core/src/protocol/validator.ts
@@ -58,8 +58,8 @@ scheme.SerializedValue = tObject({
   d: tOptional(tString),
   u: tOptional(tString),
   bi: tOptional(tString),
-  m: tOptional(tString),
-  se: tOptional(tString),
+  m: tOptional(tType('SerializedValue')),
+  se: tOptional(tType('SerializedValue')),
   r: tOptional(tObject({
     p: tString,
     f: tString,
diff --git a/packages/playwright-core/src/server/isomorphic/utilityScriptSerializers.ts b/packages/playwright-core/src/server/isomorphic/utilityScriptSerializers.ts
index 9be8d04a3..90e4b913e 100644
--- a/packages/playwright-core/src/server/isomorphic/utilityScriptSerializers.ts
+++ b/packages/playwright-core/src/server/isomorphic/utilityScriptSerializers.ts
@@ -20,8 +20,8 @@ export type SerializedValue =
     { d: string } |
     { u: string } |
     { bi: string } |
-    { m: string } |
-    { se: string } |
+    { m: SerializedValue } |
+    { se: SerializedValue } |
     { r: { p: string, f: string} } |
     { a: SerializedValue[], id: number } |
     { o: { k: string, v: SerializedValue }[], id: number } |
@@ -105,9 +105,9 @@ export function source() {
       if ('bi' in value)
         return BigInt(value.bi);
       if ('m' in value)
-        return new Map(Object.entries(JSON.parse(value.m)));
+        return new Map(Object.entries(parseEvaluationResultValue(value.m)));
       if ('se' in value)
-        return new Set(JSON.parse(value.se));
+        return new Set(parseEvaluationResultValue(value.se));
       if ('r' in value)
         return new RegExp(value.r.p, value.r.f);
       if ('a' in value) {
@@ -178,9 +178,9 @@ export function source() {
       return { bi: value.toString() };
 
     if (isMap(value))
-      return { m: JSON.stringify(Object.fromEntries(value)) };
+      return { m: serialize(Object.fromEntries(value), handleSerializer, visitorInfo) };
     if (isSet(value))
-      return { se: JSON.stringify(Array.from(value)) };
+      return { se: serialize(Array.from(value), handleSerializer, visitorInfo) };
 
     if (isError(value)) {
       const error = value;
diff --git a/packages/protocol/src/channels.ts b/packages/protocol/src/channels.ts
index c57877321..9e7abcbd1 100644
--- a/packages/protocol/src/channels.ts
+++ b/packages/protocol/src/channels.ts
@@ -180,8 +180,8 @@ export type SerializedValue = {
   d?: string,
   u?: string,
   bi?: string,
-  m?: string,
-  se?: string,
+  m?: SerializedValue,
+  se?: SerializedValue,
   r?: {
     p: string,
     f: string,
diff --git a/packages/protocol/src/protocol.yml b/packages/protocol/src/protocol.yml
index bf7cd020e..72f4bf809 100644
--- a/packages/protocol/src/protocol.yml
+++ b/packages/protocol/src/protocol.yml
@@ -83,9 +83,9 @@ SerializedValue:
     # String representation of BigInt.
     bi: string?
     # String representation of Map.
-    m: string?
+    m: SerializedValue?
     # String representation of Set.
-    se: string?
+    se: SerializedValue?
     # Regular expression pattern and flags.
     r:
       type: object?
```

**Comment by sand4rt** [comment]
> I'd do something like this instead which follows our existing serialisation infra, which would also work with circular objects, complex data types inside Map etc. This would be worth adding some tests.
> 
> 
> 
> ```diff
> 
> diff --git a/packages/playwright-core/src/protocol/serializers.ts b/packages/playwright-core/src/protocol/serializers.ts
> 
> index 3e1f99a41..813ab22cb 100644
> 
> --- a/packages/playwright-core/src/protocol/serializers.ts
> 
> +++ b/packages/playwright-core/src/protocol/serializers.ts
> 
> @@ -76,9 +76,9 @@ function innerParseSerializedValue(value: SerializedValue, handles: any[] | unde
> 
>    if (value.r !== undefined)
> 
>      return new RegExp(value.r.p, value.r.f);
> 
>    if (value.m !== undefined)
> 
> -    return new Map(Object.entries(JSON.parse(value.m)));
> 
> +    return new Map(Object.entries(innerParseSerializedValue(value.m, handles, refs)));
> 
>    if (value.se !== undefined)
> 
> -    return new Set(JSON.parse(value.se));
> 
> +    return new Set(innerParseSerializedValue(value.se, handles, refs));
> 
>  
> 
>    if (value.a !== undefined) {
> 
>      const result: any[] = [];
> 
> @@ -150,9 +150,9 @@ function innerSerializeValue(value: any, handleSerializer: (value: any) => Handl
> 
>      return { s: `${error.name}: ${error.message}\n${error.stack}` };
> 
>    }
> 
>    if (isMap(value))
> 
> -    return { m: JSON.stringify(Object.fromEntries(value)) };
> 
> +    return { m: innerSerializeValue(Object.fromEntries(value), handleSerializer, visitorInfo) };
> 
>    if (isSet(value))
> 
> -    return { se: JSON.stringify(Array.from(value)) };
> 
> +    return { se: innerSerializeValue(Array.from(value), handleSerializer, visitorInfo) };
> 
>    if (isDate(value))
> 
>      return { d: value.toJSON() };
> 
>    if (isURL(value))
> 
> diff --git a/packages/playwright-core/src/protocol/validator.ts b/packages/playwright-core/src/protocol/validator.ts
> 
> index 71f1ffbd0..c05a349a5 100644
> 
> --- a/packages/playwright-core/src/protocol/validator.ts
> 
> +++ b/packages/playwright-core/src/protocol/validator.ts
> 
> @@ -58,8 +58,8 @@ scheme.SerializedValue = tObject({
> 
>    d: tOptional(tString),
> 
>    u: tOptional(tString),
> 
>    bi: tOptional(tString),
> 
> -  m: tOptional(tString),
> 
> -  se: tOptional(tString),
> 
> +  m: tOptional(tType('SerializedValue')),
> 
> +  se: tOptional(tType('SerializedValue')),
> 
>    r: tOptional(tObject({
> 
>      p: tString,
> 
>      f: tString,
> 
> diff --git a/packages/playwright-core/src/server/isomorphic/utilityScriptSerializers.ts b/packages/playwright-core/src/server/isomorphic/utilityScriptSerializers.ts
> 
> index 9be8d04a3..90e4b913e 100644
> 
> --- a/packages/playwright-core/src/server/isomorphic/utilityScriptSerializers.ts
> 
> +++ b/packages/playwright-core/src/server/isomorphic/utilityScriptSerializers.ts
> 
> @@ -20,8 +20,8 @@ export type SerializedValue =
> 
>      { d: string } |
> 
>      { u: string } |
> 
>      { bi: string } |
> 
> -    { m: string } |
> 
> -    { se: string } |
> 
> +    { m: SerializedValue } |
> 
> +    { se: SerializedValue } |
> 
>      { r: { p: string, f: string} } |
> 
>      { a: SerializedValue[], id: number } |
> 
>      { o: { k: string, v: SerializedValue }[], id: number } |
> 
> @@ -105,9 +105,9 @@ export function source() {
> 
>        if ('bi' in value)
> 
>          return BigInt(value.bi);
> 
>        if ('m' in value)
> 
> -        return new Map(Object.entries(JSON.parse(value.m)));
> 
> +        return new Map(Object.entries(parseEvaluationResultValue(value.m)));
> 
>        if ('se' in value)
> 
> -        return new Set(JSON.parse(value.se));
> 
> +        return new Set(parseEvaluationResultValue(value.se));
> 
>        if ('r' in value)
> 
>          return new RegExp(value.r.p, value.r.f);
> 
>        if ('a' in value) {
> 
> @@ -178,9 +178,9 @@ export function source() {
> 
>        return { bi: value.toString() };
> 
>  
> 
>      if (isMap(value))
> 
> -      return { m: JSON.stringify(Object.fromEntries(value)) };
> 
> +      return { m: serialize(Object.fromEntries(value), handleSerializer, visitorInfo) };
> 
>      if (isSet(value))
> 
> -      return { se: JSON.stringify(Array.from(value)) };
> 
> +      return { se: serialize(Array.from(value), handleSerializer, visitorInfo) };
> 
>  
> 
>      if (isError(value)) {
> 
>        const error = value;
> 
> diff --git a/packages/protocol/src/channels.ts b/packages/protocol/src/channels.ts
> 
> index c57877321..9e7abcbd1 100644
> 
> --- a/packages/protocol/src/channels.ts
> 
> +++ b/packages/protocol/src/channels.ts
> 
> @@ -180,8 +180,8 @@ export type SerializedValue = {
> 
>    d?: string,
> 
>    u?: string,
> 
>    bi?: string,
> 
> -  m?: string,
> 
> -  se?: string,
> 
> +  m?: SerializedValue,
> 
> +  se?: SerializedValue,
> 
>    r?: {
> 
>      p: string,
> 
>      f: string,
> 
> diff --git a/packages/protocol/src/protocol.yml b/packages/protocol/src/protocol.yml
> 
> index bf7cd020e..72f4bf809 100644
> 
> --- a/packages/protocol/src/protocol.yml
> 
> +++ b/packages/protocol/src/protocol.yml
> 
> @@ -83,9 +83,9 @@ SerializedValue:
> 
>      # String representation of BigInt.
> 
>      bi: string?
> 
>      # String representation of Map.
> 
> -    m: string?
> 
> +    m: SerializedValue?
> 
>      # String representation of Set.
> 
> -    se: string?
> 
> +    se: SerializedValue?
> 
>      # Regular expression pattern and flags.
> 
>      r:
> 
>        type: object?
> 
> ```

Thanks for the pointers. I've updated the PR.

File: packages/playwright-core/src/server/isomorphic/utilityScriptSerializers.ts [hunk 1]
```diff
@@ -163,6 +177,11 @@ export function source() {
     if (typeof value === 'bigint')
       return { bi: value.toString() };
 
+    if (isMap(value))
+      return { m: serialize(Object.fromEntries(value), handleSerializer, visitorInfo) };
```
Comment (dgozman) [inline packages/playwright-core/src/server/isomorphic/utilityScriptSerializers.ts:181]: I believe  this turns non-string keys into strings. I think we should do manual serialization: array of values for set, array `[key1, value1, key2, value2, ...]` for maps.

Comment (sand4rt) [reply packages/playwright-core/src/server/isomorphic/utilityScriptSerializers.ts]: resolved in: https://github.com/microsoft/playwright/pull/26730/commits/01a3f7180c7c014d3f09790c5136f78b9c154033

File: packages/protocol/src/protocol.yml [hunk 2]
```diff
@@ -82,6 +82,10 @@ SerializedValue:
     u: string?
     # String representation of BigInt.
     bi: string?
+    # String representation of Map.
```
Comment (dgozman) [inline packages/protocol/src/protocol.yml:85]:
```suggestion
    # JS representation of Map: [[key1, value1], [key2, value2], ...]
```

File: packages/protocol/src/protocol.yml [hunk 3]
```diff
@@ -82,6 +82,10 @@ SerializedValue:
...
+    m: SerializedValue?
+    # String representation of Set.
```
Comment (dgozman) [inline packages/protocol/src/protocol.yml:87]:
```suggestion
    # JS representation of Set: [item1, item2, ...]
```

**Review by dgozman — COMMENTED** [review]
Looks great, just some nits.

**Review by dgozman — APPROVED** [review]

**Pull request merged**